Make shallow cloning more durable

timrchavez opened the pull request:
If a git-resource is configured to use shallow cloning and new commits
are added to the repository that are not considered new versions of the
resource (e.g. via a path filter excludes it), a `get` will shallow
clone HEAD which will not be equal to the version ref. This will result
in breakage.

pivotal-issuemaster commented:
@timrchavez Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/git-resource&pullRequestId=167)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/git-resource&pullRequestId=167) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

timrchavez commented:
Assuming I have the test correct, I do not believe the suggested change for #158 will work. The error I'm getting is:

```
running it_can_shallow_clone_ref_with_new_commits...
  Switched to a new branch 'bogus'
  Switched to branch 'master'
  Cloning into '/tmp/git-tests.dlHJcd/git-tests.JFhcGp/destination'...
  error: Server does not allow request for unadvertised object 57b3b83c0199e8d73a1d1921785f2f73b0eb5ddf
The command '/bin/sh -c /tests/all.sh' returned a non-zero code: 1
```

bhcleek commented:
On a shallow clone, I am able to duplicate the error you're seeing. It appears, though, that setting `uploadpack.allowReachableSHA1InWant` on the server will allow an arbitrary SHA to be retrieved.

RE: https://git-scm.com/docs/git-fetch-pack

timrchavez commented:
I use a good ol `until` loop to progressively fetch deeper and deeper until I can successfully check out the ref.

vito commented on assets/in:
```diff
@@ -63,7 +63,11 @@ git clone --single-branch $depthflag $uri $branchflag $destination
 
 git fetch origin refs/notes/*:refs/notes/*
-git checkout -q $ref
+until `git checkout -q $ref`; do
```
This will result in an infinite loop if `$ref` is no longer available anywhere (i.e. someone `push -f`ed).

timrchavez replied:
Hum. Good catch!

vito commented on assets/in:
```diff
@@ -63,7 +63,11 @@ git clone --single-branch $depthflag $uri $branchflag $destination
 git fetch origin refs/notes/*:refs/notes/*
-git checkout -q $ref
+until `git checkout -q $ref`; do
+    deepen=$((deepen+$depth))
```
This only makes sense to attempt if `depth` is given (it defaults to 0).

timrchavez replied:
:+1:

timrchavez commented:
@vito Yep. I should have time. I'm not sure about @bhcleek's approach honestly because it requires a configuration change to the GitHub installation itself _and_ based on the documentation it's very resource intensive. I'm open to other suggestions, but I'm not creative enough to think of anything better than a loop which fetches deeper and deeper until it can find its ref.

vito commented:
@timrchavez I'm fine with it fetching all the way back to the first commit in the `push -f` case and then giving up, just as long as it's an infinite loop.

vito commented:
merged #205
